File: 백준/Gold/2252. 줄 세우기/줄 세우기.cc
#include <bits/stdc++.h>
using namespace std;

int n, m;
int deg[32005];
vector<int> adj[32005];
vector<int> result;

void topologicalSort() {
	queue<int> q;
	for (int i = 1; i <= n; i++)
		if (deg[i] == 0) q.push(i);

	while (!q.empty()) {
		int cur = q.front(); q.pop();
		result.push_back(cur);
		for (int nxt : adj[cur]) {
			deg[nxt]--;
			if (deg[nxt] == 0) q.push(nxt);
		}
	}

	for (int i : result) cout << i << ' ';
}

int main() {
	ios::sync_with_stdio(false);
	cin.tie(NULL);

	cin >> n >> m;
	while (m--) {
		int a, b;
		cin >> a >> b;
		adj[a].push_back(b);
		deg[b]++;
	}

	topologicalSort();
	return 0;
}

File: 백준/Gold/2252. 줄 세우기/줄 세우기.py
from collections import deque
import sys
input = sys.stdin.readline

def topological_sort(n, adj, deg):
    result = []
    q = deque()

    for i in range(1, n + 1):
        if deg[i] == 0:
            q.append(i)

    while q:
        cur = q.popleft()
        result.append(cur)
        for nxt in adj[cur]:
            deg[nxt] -= 1
            if deg[nxt] == 0:
                q.append(nxt)

    if len(result) != n:
        print("Cycle exists!")
    else:
        print(*result)

n, m = map(int, input().split())

adj = {i: [] for i in range(1, n + 1)}
deg = {i: 0 for i in range(1, n + 1)}

for _ in range(m):
    a, b = map(int, input().split())
    adj[a].append(b)
    deg[b] += 1

topological_sort(n, adj, deg)